File: src/bytes_io.c
// Binary TCP: List<U32> of 0..255, not UTF-8 strings.
// ================================================

static u32 cinder_u32(Term t) {
  return (u32)(u64)t;
}

static Term bytes_recv_pack(Env e, IoWork* w) {
  Term r;
  if (w->code) {
    r = io_fail(e, w->code, NULL);
  } else {
    Term xs = term_pak(CID_NIL, 0);
    for (u64 i = w->size; i > 0; i -= 1) {
      xs = io_node(e, CID_CON, ((uint8_t*)w->data)[i - 1], xs, IO_HOTS & 16);
    }
    r = io_done(e, xs);
  }
  free(w->data);
  return io_tup(e, io_hand(w->hand), r);
}

static Term bytes_recv_more(Env e, IoWork* w) {
  int fd  = (int)w->hand;
  w->size = io_sys_end(w, recv(fd, w->data, (size_t)w->made, 0));
  return w->code == EAGAIN ? io_wait_on(w, fd, POLLIN, bytes_recv_more)
    : bytes_recv_pack(e, w);
}

Term bytes_recv_run(Env e, Term* f, IoWork* w) {
  w->hand = (intptr_t)io_hand_v(f[0]);
  w->made = f[1] < INT32_MAX ? (intptr_t)f[1] : INT32_MAX;
  if (w->made > 65536) {
    w->made = 65536;
  }
  w->data = io_mem(malloc((size_t)w->made + 1));
  return bytes_recv_more(e, w);
}

static Term bytes_send_more(Env e, IoWork* w) {
  int fd = (int)w->hand;
  while (w->code == 0 && (u64)w->made < w->size) {
    ssize_t n = send(fd, (char*)w->data + w->made, w->size - (u64)w->made, 0);
    if (n < 0 && errno == EAGAIN) {
      return io_wait_on(w, fd, POLLOUT, bytes_send_more);
    }
    w->made += io_sys_end(w, n);
  }
  Term r = w->code != 0 ? io_fail(e, w->code, NULL)
    : io_done(e, term_pak(CID_UNIT, 0));
  free(w->data);
  return io_tup(e, io_hand(w->hand), r);
}

Term bytes_send_run(Env e, Term* f, IoWork* w) {
  u64 cap = 256;
  u64 n = 0;
  char* data = io_mem(malloc(cap));
  Term s = f[1];
  while (term_aux(s) == CID_CON) {
    Term fb[2];
    spare_free(e, cls_fit(2), ctr_take(e, s, 2, fb));
    if (n == cap) {
      cap *= 2;
      data = io_mem(realloc(data, cap));
    }
    data[n] = (uint8_t)cinder_u32(fb[0]);
    n += 1;
    s = fb[1];
  }
  w->hand = (intptr_t)io_hand_v(f[0]);
  w->data = data;
  w->size = n;
  w->made = 0;
  w->code = 0;
  return bytes_send_more(e, w);
}

static void __attribute__((constructor)) bytes_io_use(void) {
  io_eff(CID_BYTES_RECV, bytes_recv_run, IO_READ);
  io_eff(CID_BYTES_SEND, bytes_send_run, 0);
}

File: src/bytes_io.js
// Binary TCP: List of 0..255, not UTF-8 strings.

function bytes_list(u8, n) {
  let xs = { $: "Nil" };
  for (let i = n - 1; i >= 0; i -= 1) {
    xs = { $: "Con", head: u8[i], tail: xs };
  }
  return xs;
}

function bytes_flat(xs) {
  const out = [];
  let cur = xs;
  while (cur && cur.$ === "Con") {
    out.push(Number(cur.head) & 255);
    cur = cur.tail;
  }
  return Uint8Array.from(out);
}

function bytes_recv(socket, max, k) {
  const sys = io_sys();
  const fd = socket;
  const cap = Math.max(1, Math.min(Number(max), 65536));
  const b = new Uint8Array(cap);
  const again = sys.mac ? 35 : 11;
  const go = () => {
    const n = Number(sys.recv(fd, sys.ptr(b), cap, 0));
    if (n < 0) {
      const code = sys.errno();
      if (code === again) {
        io_park_on(fd, false, k, go);
        return undefined;
      }
      return io_tup(socket, io_fail(code));
    }
    return io_tup(socket, io_done(bytes_list(b, n)));
  };
  return go();
}

function bytes_recv_need() {
  return { read: true };
}

function bytes_send(socket, data, k) {
  const sys = io_sys();
  const fd = socket;
  const b = bytes_flat(data);
  const again = sys.mac ? 35 : 11;
  const go = (at) => {
    while (at < b.length) {
      const part = b.subarray(at);
      const n = Number(sys.send(fd, sys.ptr(part), part.length, 0));
      if (n < 0) {
        const code = sys.errno();
        if (code === again) {
          io_park_on(fd, true, k, () => go(at));
          return undefined;
        }
        return io_tup(socket, io_fail(code));
      }
      at += n;
    }
    return io_tup(socket, io_done({ $: "Unit" }));
  };
  return go(0);
}
